framework/src/modules/dpos/transaction_assets/pom_transaction_asset.ts

```diff
@@ -0,0 +1,264 @@
+			throw new Error(
+				`Difference between header1.height and current height must be less than ${MAX_PUNISHABLE_BLOCK_HEIGHT_DIFFERENCE.toString()}.`,
+			);
+		}
+
+		if (Math.abs(asset.header2.height - currentHeight) >= MAX_PUNISHABLE_BLOCK_HEIGHT_DIFFERENCE) {
+			throw new Error(
+				`Difference between header2.height and current height must be less than ${MAX_PUNISHABLE_BLOCK_HEIGHT_DIFFERENCE.toString()}.`,
+			);
+		}
+
+		/*
+			Check if delegate is eligible to be punished
+		*/
+		const delegateAddress = getAddressFromPublicKey(asset.header1.generatorPublicKey);
+		const delegateAccount = await store.account.get<DPOSAccountProps>(delegateAddress);
+
+		if (delegateAccount.dpos.delegate.username === '') {
+			throw new Error('Account is not a delegate');
+		}
+
+		if (delegateAccount.dpos.delegate.isBanned) {
+			throw new Error('Cannot apply proof-of-misbehavior. Delegate is banned.');
+		}
+
+		if (
+			getPunishmentPeriod(
+				delegateAccount,
+				delegateAccount,
+				store.chain.lastBlockHeaders[0].height,
+			) > 0
+		) {
+			throw new Error('Cannot apply proof-of-misbehavior. Delegate is already punished. ');
+		}
+
+		/*
+			Check block signatures validity
+		*/
+
+		const blockHeader1Bytes = Buffer.concat([
+			networkIdentifier,
+			getBlockHeaderBytes(asset.header1),
+		]);
+		if (
+			!validateSignature(
+				asset.header1.generatorPublicKey,
+				asset.header1.signature,
+				blockHeader1Bytes,
+			)
+		) {
+			throw new Error('Invalid block signature for header 1.');
+		}
+
+		const blockHeader2Bytes = Buffer.concat([
+			networkIdentifier,
+			getBlockHeaderBytes(asset.header2),
+		]);
+
+		if (
+			!validateSignature(
+				asset.header2.generatorPublicKey,
+				asset.header2.signature,
+				blockHeader2Bytes,
+			)
+		) {
+			throw new Error('Invalid block signature for header 2.');
+		}
+
+		/*
+			Update sender account
+		*/
+		const delegateAccountBalance = await reducerHandler.invoke<bigint>('token:getBalance', {
+			address: delegateAccount.address,
+		});
+
+		const reward =
+			store.chain.lastBlockReward > delegateAccountBalance
+				? delegateAccountBalance
+				: store.chain.lastBlockReward;
+
+		await reducerHandler.invoke('token:credit', { address: senderID, amount: reward });
+
+		store.account.set(senderAccount.address, senderAccount);
+
+		/*
+			Update delegate account
+		*/
+
+		// Fetch delegate account again in case sender and delegate are the same account
+		const updatedDelegateAccount = await store.account.get<DPOSAccountProps>(delegateAddress);
+
+		updatedDelegateAccount.dpos.delegate.pomHeights.push(currentHeight);
+
+		if (updatedDelegateAccount.dpos.delegate.pomHeights.length >= MAX_POM_HEIGHTS) {
+			updatedDelegateAccount.dpos.delegate.isBanned = true;
+		}
+		await reducerHandler.invoke('token:debit', {
+			address: updatedDelegateAccount.address,
+			amount: reward,
+		});
+		store.account.set(updatedDelegateAccount.address, updatedDelegateAccount);
```

Comment:
this should happen before invoking `token:debit` so that they can fetch updated account